### ISSUE TITLE
Rust Client Side SDK

### ISSUE THREAD

**Issue by markphelps** [flipt-io/flipt-client-sdks#83]
### Language

Rust :crab:

### FFI Support

- [X] I've validated that the language supports Foreign Function Interface (FFI) bindings

### Additional Context

I dont think we'll need to go through FFI since our engine is written in Rust

### Willing to Implement

- [X] I'm willing to help implement the Flipt Client SDK in this language if it's accepted. 🙌

### Willing to Review

- [X] I'm willing to review contributions to the Flipt Client SDK in this language if it's accepted.

**Comment by andrew-scott-fischer** [comment]
@markphelps What prevented this from being worked on? are you open to PRs to add it?

**Comment by markphelps** [comment]
> [@markphelps](https://github.com/markphelps) What prevented this from being worked on? are you open to PRs to add it?

Only our lack of Rust library experience. Def open to PRs!!

**Comment by andrew-scott-fischer** [comment]
I'll propose this as a starting point. Do you have any concerns with assuming a `tokio` runtime and using `rustls` rather than `native-tls`?

I considered making it runtime independent and allowing clients to configure how they wanted to do mtls, but I thought it might be a bit over engineered.

I thought it reasonable to make the system async by default, but provide a `blocking` feature so sync systems could still leverage the client, and that didn't seem like much of a lift.

I'd like some feedback on what constraints you think are important before I begin proposing PRs.

---

# Native Rust Client SDK Design

## Goal

Add a publishable, user-facing Rust client to `flipt-client-sdks` with a simple Tokio-first API, while preserving the existing FFI and WASM SDKs.

## Primary decision

Use Tokio and Reqwest directly in v1. Do not build a runtime-neutral abstraction in the first implementation.

The normal async path assumes the application is running a Tokio runtime:

```rust
let client = Client::builder()
    .url("https://flipt.example.com")
    .namespace("default")
    .build()
    .await?;

let result = client.evaluate_boolean(&request)?;
```

The client does not construct a hidden runtime or call `block_on` in the async path. It uses the caller's Tokio runtime for background updates.

The `blocking` feature provides a synchronous facade that explicitly owns a Tokio runtime.

## Architecture

```mermaid
flowchart TD
    Eval["flipt-evaluation<br/>Evaluation and models"]
    Core["flipt-client-core<br/>Tokio transport and state"]
    Client["flipt-client-rust<br/>Public Rust API"]
    Blocking["BlockingClient<br/>Optional feature"]
    FFI["flipt-engine-ffi<br/>Existing FFI adapter"]
    AppRuntime["Application Tokio runtime"]
    BlockingRuntime["Blocking-owned Tokio runtime"]
    FFIRuntime["Existing FFI runtime bridge"]

    Eval --> Core
    Eval --> Client
    Eval --> FFI
    Core --> Client
    Core --> FFI
    Client --> Blocking
    Client --> AppRuntime
    Blocking --> BlockingRuntime
    FFI --> FFIRuntime
```

### `flipt-evaluation`

A publishable library containing:

- Evaluation algorithms
- Evaluation requests and responses
- Normalized snapshot models
- Evaluation errors and reasons

Before publication, change the Rust library identifier from `fliptevaluation` to the idiomatic `flipt_evaluation` and update all workspace imports:

```toml
[package]
name = "flipt-evaluation"

[lib]
name = "flipt_evaluation"
```

This crate remains synchronous and has no Tokio or HTTP dependency.

### `flipt-client-core`

A publishable, Tokio-based client engine containing:

- Reqwest HTTP transport
- Rustls TLS configuration
- Initial snapshot fetch
- Polling and streaming updates
- Retry and timeout handling
- Authentication and environment headers
- Snapshot state ownership and replacement
- One adapter-driven update future

The core may use Tokio APIs in v1, but it must not create a Tokio runtime or call `block_on`. The caller or adapter owns scheduling.

The update loop must be consolidated rather than moving `HTTPFetcher::start` unchanged. The core exposes one future:

```text
run_updates()
  wait
  fetch
  apply snapshot
  repeat
```

The async Rust client schedules that future with `tokio::spawn`. The FFI adapter schedules it through its existing runtime bridge.

```mermaid
flowchart LR
    Transport["Reqwest transport"] --> Updates["Core run_updates future"]
    Updates --> State["Shared snapshot state"]
    State --> Client["Async Client"]
    State --> FFI["FFI Engine"]
    Client -->|"tokio spawn"| App["Application runtime"]
    FFI -->|"Handle spawn"| Legacy["FFI runtime"]
```

### `flipt-client-rust`

The public native Rust API should provide:

- `ClientBuilder`
- `Client`
- Typed options and authentication
- Boolean evaluation
- Variant evaluation
- Batch evaluation
- List flags
- Snapshot access/import
- Async polling/streaming lifecycle
- Async shutdown

`Client` should be cloneable over shared state:

```rust
#[derive(Clone)]
pub struct Client {
    inner: Arc<ClientInner>,
}
```

`close` should take `&self` so every clone observes the same shutdown state. It should be idempotent:

```rust
pub async fn close(&self) -> Result<(), Error>;
```

The shared inner state should contain a cancellation signal and the update task. The first caller to close awaits the task; later callers observe that shutdown has already started or completed.

`Drop` must never block or create a runtime. When the final `Arc` is dropped without an explicit close, it should cancel and abort the update task as a best-effort fallback.

### `BlockingClient`

Enable the synchronous facade with the `blocking` feature:

```rust
let client = BlockingClient::builder()
    .url("https://flipt.example.com")
    .build()?;
```

`BlockingClient` owns a dedicated multi-thread Tokio runtime so polling and streaming continue between synchronous calls:

```rust
pub struct BlockingClient {
    client: Option<Client>,
    runtime: Option<tokio::runtime::Runtime>,
}
```

The `client` field is declared first and the `runtime` field last so the client is torn down before the runtime. `close(self)` takes both options, runs the shared async shutdown on the owned runtime, and then releases the client and runtime in that order. `Drop` cancels and aborts without calling `block_on`.

The blocking facade is intended for synchronous threads. Applications already running async code should use `Client` instead.

## Feature and dependency model

The first implementation keeps the feature set small:

```toml
[features]
default = ["rustls-tls"]
rustls-tls = []
blocking = []
```

Tokio and Reqwest are normal v1 dependencies of the core. `blocking` enables the synchronous API; it does not select a second runtime.

Use Rustls as the only v1 TLS backend. The existing PEM-based Rustls identity path can support mutual TLS without introducing a second configuration model.

## Implementation plan

### 1. Stabilize the evaluation crate

Modify:

- `Cargo.toml`
- `flipt-evaluation/Cargo.toml`
- `flipt-evaluation/src/lib.rs`
- Existing Rust imports in engine crates

Actions:

- Rename the library identifier to `flipt_evaluation`.
- Make package metadata complete for publication.
- Add explicit versions to published path dependencies.
- Review public model and error types for semver stability.
- Keep evaluation behavior unchanged.

### 2. Extract the Tokio client core

Create:

- `flipt-client-core/Cargo.toml`
- `flipt-client-core/src/lib.rs`
- `flipt-client-core/src/http.rs`
- `flipt-client-core/src/tls.rs`
- `flipt-client-core/src/error.rs`
- `flipt-client-core/src/lifecycle.rs`

Modify:

- `flipt-engine-ffi/Cargo.toml`
- `flipt-engine-ffi/src/lib.rs`
- `flipt-engine-ffi/src/http.rs`
- `flipt-engine-ffi/src/tls.rs`

Actions:

- Move shared Reqwest, polling, streaming, authentication, TLS, retry, and snapshot-update logic into the core.
- Replace nested fetcher/engine spawning with one `run_updates` future.
- Keep runtime construction, `block_on`, and FFI shutdown bridging in `flipt-engine-ffi`.
- Preserve all existing FFI symbols and behavior.

### 3. Add the native Rust client

Create:

- `flipt-client-rust/Cargo.toml`
- `flipt-client-rust/src/lib.rs`
- `flipt-client-rust/src/client.rs`
- `flipt-client-rust/src/options.rs`
- `flipt-client-rust/src/error.rs`
- `flipt-client-rust/README.md`
- `flipt-client-rust/tests/*`

Actions:

- Add the builder-based happy path with no mandatory transport injection.
- Wrap the core in `Arc`-backed cloneable client state.
- Spawn one update task on the caller's Tokio runtime.
- Implement synchronous local evaluation methods.
- Implement idempotent `close(&self).await`.
- Implement non-blocking cancellation/abort on final drop.

### 4. Add blocking support

Actions:

- Gate `BlockingClient` behind `blocking`.
- Give it an owned multi-thread Tokio runtime.
- Delegate all behavior to the async client/core.
- Add blocking construction, evaluation, snapshot, update, and close tests.
- Document that blocking clients must not be used from async worker threads.

### 5. Verify behavior and compatibility

Add tests for:

- Boolean, variant, batch, and list-flags evaluation
- Snapshot import/export
- Initial fetch and polling updates
- Streaming updates
- Authentication and environment headers
- Retry and timeout behavior
- Rustls CA and PEM mTLS configuration
- Async close and final-drop cancellation
- Blocking close and runtime teardown
- FFI behavior after core extraction

Run focused Rust tests, formatting, Clippy, and package validation for each publishable crate. Add a test or build check proving the update loop has one scheduler-owned task rather than nested spawning.

### 6. Wire project automation and documentation

Modify:

- Root `Cargo.toml`
- Root `.mise.toml`
- Rust CI workflows
- `release/release_cli.py`
- `release/sdks/*`
- `.github/dependabot.yml`
- Root `README.md`
- `ARCHITECTURE.md`
- `CONTRIBUTING.md`

Add:

- Workspace members for all three crates
- Rust client lint/test coverage
- Cargo package and publication checks
- crates.io release handling
- Rust client documentation and examples
- Rust integration-test registration where appropriate

Move Rust from Planned to Released only after publication, integration tests, and FFI compatibility checks succeed.

## Acceptance criteria

- A downstream Cargo project can depend on `flipt-client-rust`.
- `Client::builder().url(...).build().await?` works with the default Rustls transport.
- Async users run the client on their Tokio runtime without hidden runtime creation.
- `BlockingClient` works with the `blocking` feature and owns its runtime explicitly.
- Polling and streaming use one update task per client.
- Existing FFI and WASM SDK behavior remains unchanged.
- Rustls PEM mutual TLS works in v1.
- CI, release automation, documentation, and Dependabot support the new client.

## Alternatives considered

### Runtime-neutral async support

```mermaid
flowchart LR
    Eval["Evaluation"] --> Core["Runtime-neutral core"]
    Core --> Traits["Transport and timing traits"]
    Traits --> Adapter["Optional runtime adapter"]
    Adapter --> Runtime["Caller-selected runtime"]
```

A runtime-neutral design would remove Tokio and Reqwest from the default core, define transport plus sleep/clock traits, and provide Tokio/Reqwest as an optional adapter. The public client would either accept an adapter or expose caller-driven update futures. This supports multiple executors but adds abstraction, feature, testing, and documentation complexity. It is intentionally deferred.

### Native TLS mutual TLS

```mermaid
flowchart LR
    Client["Client"] --> Backend["TLS backend"]
    Backend --> Rustls["Rustls"]
    Rustls --> Pem["PEM identity"]
    Backend --> Native["Native TLS"]
    Native --> Pkcs12["PKCS12 identity"]
```

Supporting native-tls mTLS would require a backend-specific adapter and identity API. The existing Rustls PEM identity configuration cannot be reused directly; native-tls generally requires PKCS#12 identity material. This design would add native-tls feature wiring, conditional TLS types, PKCS#12 configuration, platform-specific CI, and a larger compatibility matrix. v1 uses Rustls PEM mTLS and defers native-tls mTLS until that API is designed.

**Comment by markphelps** [comment]
I think this design looks great. Going with Tokio makes sense especially for the initial version. Async by default + blocking feature also makes sense to me.My only feedback/question is which TLS root source do we use? 

 - rustls-tls-native-roots: use OS trust roots.
 - rustls-tls-webpki-roots: bundle Mozilla roots. This is more predictable for containers and static binaries.

**Comment by markphelps** [comment]
Also thank you for the very detailed design proposal @andrew-scott-fischer !
